src/NHibernate/Impl/SessionImpl.cs

```diff
@@ -347,6 +354,9 @@ private void Cleanup()
 		{
 			using (new SessionIdLoggingContext(SessionId))
 			{
+				// Let the after tran clear that if we are still in an active system transaction.
+				if (TransactionContext?.IsInActiveTransaction == true)
+					return;
 				persistenceContext.Clear();
```

Comment:
Delaying that in case of ongoing transaction does look required currently, but since `TransactionCompletion` does act on  `persistenceContext`, it looks more consistent to me.